staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -654,14 +634,44 @@ func transformErrorToEvent(err error) *watch.Event {
 	}
 }
 
+// sendError synchronously puts an error event into resultChan and
+// trigger cancelling all goroutines.
 func (wc *watchChan) sendError(err error) {
+	if isCancelError(err) {
+		return
+	}
+	errResult := transformErrorToEvent(err)
+	if errResult != nil {
+		// error result is guaranteed to be received by user before closing ResultChan.
+		select {
+		case wc.resultChan <- *errResult:
+		case <-wc.ctx.Done(): // user has given up all results
+		}
+	}
+
+	// We use wc.ctx to reap all goroutines. Under whatever condition, we should stop them all.
+	// It's fine to double cancel.
+	wc.cancel()
```

Comment:
noticed this is a new addition, more than just a refactor... trying to trace everything this connects to...